### PR TITLE
move sql tests from `context.rs` to corresponding test files in `tests/sql`

### LINKED ISSUE #2328: move sql tests in `context.rs` to corresponding test files in `datafustion/core/tests/sql`
`datafusion/core/src/execution/context.rs` has many unmanaged test cases. 
We can move them to corresponding test files in `datafustion/core/tests/sql` for better code structure.

### PR DESCRIPTION
**Pull request by WinkerDu**
# Which issue does this PR close?

Closes #2328 .
Closes https://github.com/apache/arrow-datafusion/issues/743

 # Rationale for this change

`datafusion/core/src/execution/context.rs` has many unmanaged test cases.
We can move them to corresponding test files in `datafustion/core/tests/sql` for better code structure.

# What changes are included in this PR?

move sql tests from `context.rs` to corresponding test files in `tests/sql`

# Are there any user-facing changes?

No.

### CODE REVIEW

**Review by andygrove — APPROVED** [review]
I haven't reviewed this line by line and assume this is just moving the tests. I like this approach because it will make it easier to run these SQL tests against other contexts in the future (e.g. Ballista) if we ever decide to do that.

**Comment by WinkerDu** [comment]
Thank you ❤️  @andygrove, I think there is nothing special between test cases in `context.rs` and those in `tests/sql` since they  both use `SessionContext`, maybe it's ok to move.

@alamb @yjshen @xudong963  PTAL, thank you.

File: datafusion/core/src/execution/context.rs [hunk 1]
```diff
@@ -1663,783 +1655,48 @@ mod tests {
...
-            let val_array = Arc::new(val_array);
-
-            let schema = Arc::new(Schema::new(vec![
-                Field::new("str", str_array.data_type().clone(), false),
-                Field::new("val", val_array.data_type().clone(), false),
-            ]));
-
-            let batch =
-                RecordBatch::try_new(schema.clone(), vec![str_array, val_array]).unwrap();
+    async fn register_deregister() -> Result<()> {
```
Comment (alamb) [inline datafusion/core/src/execution/context.rs:1688]: 👍  for keeping the variable tests (and others that are related to `SessionContext` in the same file

File: datafusion/core/tests/sql/functions.rs [hunk 2]
```diff
@@ -357,3 +357,240 @@ async fn coalesce_mul_with_default_value() -> Result<()> {
     assert_batches_eq!(expected, &actual);
     Ok(())
 }
+
+#[tokio::test]
+async fn count_basic() -> Result<()> {
```
Comment (alamb) [inline datafusion/core/tests/sql/functions.rs:362]: As `COUNT` and `AVG` are aggregate functions, these tests probably belong in aggregates.rs or something similar -- we can move them as a follow on PR

**Review by alamb — APPROVED** [review]
Thank you so much @WinkerDu -- I think this completes a long standing goal to move the sql tests out of context.rs: https://github.com/apache/arrow-datafusion/issues/743 👍  Very nice.

File: datafusion/core/tests/sql/functions.rs (hunk 2 above)
Comment (alamb) [reply datafusion/core/tests/sql/functions.rs:362]: (I am working on this)

Comment (alamb) [reply datafusion/core/tests/sql/functions.rs:362]: https://github.com/apache/arrow-datafusion/pull/2333#

**Pull request merged**